File: bindings/python/kevy/_url.py
"""Connect-URL parsing and the process-global embedded registry (§1.1–§1.3).

A single ``connect(url)`` chooses the backend from the scheme. Embedded
targets (``mem://<name>`` / ``file://<path>``) share one backing Store per
URL through a process-global, refcounted registry so a publisher and a
consumer opened on the same URL find each other (§1.3). Anonymous ``mem://``
is always isolated.
"""

from __future__ import annotations

import os
import threading
from dataclasses import dataclass
from enum import Enum, auto
from typing import Optional, Tuple

from ._errors import InvalidInputError, UnsupportedError


class TargetKind(Enum):
    MEM_ANON = auto()  # mem:// — fresh, never shared
    MEM_NAMED = auto()  # mem://<name> — shared by name
    FILE = auto()  # file://<path> — shared + persistent
    REMOTE = auto()  # kevy:// / redis:// / tcp://


@dataclass
class Target:
    kind: TargetKind
    url: str
    name: str = ""  # mem://<name>
    path: str = ""  # file://<canonical-path>
    host: str = ""  # remote host
    port: int = 6379  # remote port (default 6379)
    db: Optional[int] = None  # remote /db (None for tcp://)

    def registry_key(self) -> str:
        if self.kind is TargetKind.MEM_NAMED:
            return "mem://" + self.name
        if self.kind is TargetKind.FILE:
            return "file://" + self.path
        return ""


def parse_connect_url(url: str) -> Target:
    """Resolve a connect URL to a Target, rejecting TLS/AUTH and unknown
    schemes before any I/O (§1.1)."""
    if "://" not in url:
        raise InvalidInputError("URL missing '://'")
    scheme, rest = url.split("://", 1)
    if scheme == "mem":
        if rest == "":
            return Target(kind=TargetKind.MEM_ANON, url=url)
        return Target(kind=TargetKind.MEM_NAMED, url=url, name=rest)
    if scheme == "file":
        if rest == "":
            raise InvalidInputError(
                "file:// URL must include a path (e.g. file:///var/lib/myapp)"
            )
        return Target(kind=TargetKind.FILE, url=url, path=_canonical_file_path(rest))
    if scheme in ("kevy", "redis", "tcp"):
        return _parse_remote(scheme, rest, url)
    if scheme in ("rediss", "kevys"):
        raise UnsupportedError(
            "TLS schemes (rediss://, kevys://) are unsupported — kevy has no TLS"
        )
    raise InvalidInputError(f"unknown URL scheme '{scheme}://'")


def _canonical_file_path(rest: str) -> str:
    # file:///abs → /abs ; file://./rel → ./rel. Canonicalise so two URLs
    # naming the same path share a store.
    path = rest
    return os.path.abspath(path)


def _parse_remote(scheme: str, rest: str, url: str) -> Target:
    if "@" in rest:
        raise UnsupportedError(
            "userinfo (user:pass@host) is unsupported — kevy has no AUTH"
        )
    authority = rest
    path = ""
    slash = rest.find("/")
    if slash >= 0:
        authority = rest[:slash]
        path = rest[slash + 1 :]
    host, port = _parse_authority(authority)
    target = Target(kind=TargetKind.REMOTE, url=url, host=host, port=port)
    # tcp:// is raw: it never carries a db (§1.1). kevy:// / redis:// honour /N.
    if scheme != "tcp" and path != "":
        try:
            n = int(path)
        except ValueError:
            raise InvalidInputError(
                f"bad db index: '{path}' (expected a non-negative integer)"
            )
        if n < 0:
            raise InvalidInputError(f"bad db index: '{path}' (must be non-negative)")
        target.db = n
    return target


def _parse_authority(authority: str) -> Tuple[str, int]:
    host = authority
    port = 6379
    colon = authority.rfind(":")
    if colon >= 0:
        host = authority[:colon]
        try:
            port = int(authority[colon + 1 :])
        except ValueError:
            raise InvalidInputError(f"bad port: {authority[colon + 1:]}")
        if not (0 <= port <= 65535):
            raise InvalidInputError(f"bad port: {authority[colon + 1:]}")
    if host == "":
        raise InvalidInputError("empty host")
    return host, port


# --- process-global embedded registry (§1.3) ----------------------------


class _SharedStore:
    __slots__ = ("db", "refs")

    def __init__(self, db):
        self.db = db
        self.refs = 1


_registry_lock = threading.Lock()
_registry: dict = {}


def resolve_store(target: Target):
    """Open (or share) the embedded store for a target. Returns
    ``(db, key)``; ``key`` is "" for the unshared anonymous case (§1.3)."""
    from ._embedded import open_mem, open_persistent

    key = target.registry_key()
    if key:
        with _registry_lock:
            entry = _registry.get(key)
            if entry is not None:
                entry.refs += 1
                return entry.db, key

    if target.kind in (TargetKind.MEM_ANON, TargetKind.MEM_NAMED):
        db = open_mem()
    elif target.kind is TargetKind.FILE:
        db = open_persistent(target.path)
    else:
        raise InvalidInputError("resolve_store called on a non-embedded target")

    if not key:
        return db, ""

    with _registry_lock:
        entry = _registry.get(key)
        if entry is not None:  # lost a race — use theirs, close ours
            entry.refs += 1
            db.close()
            return entry.db, key
        _registry[key] = _SharedStore(db)
        return db, key


def release_store(key: str, db) -> None:
    """Drop one reference; close the store when the last handle goes."""
    if not key:
        db.close()
        return
    with _registry_lock:
        entry = _registry.get(key)
        if entry is None:
            return
        entry.refs -= 1
        if entry.refs <= 0:
            entry.db.close()
            del _registry[key]

File: bindings/cpp/src/url.cpp
#include "url.hpp"

#include <charconv>
#include <mutex>
#include <unordered_map>

#include "kevy/errors.hpp"

namespace kevy {
namespace detail {

std::string Target::registry_key() const {
  switch (kind) {
    case TargetKind::MemNamed: return "mem://" + name;
    case TargetKind::File: return "file://" + path;
    default: return "";
  }
}

namespace {

bool parse_u32(const std::string& s, uint32_t& out) {
  auto res = std::from_chars(s.data(), s.data() + s.size(), out);
  return res.ec == std::errc() && res.ptr == s.data() + s.size();
}

Target parse_remote(const std::string& scheme, const std::string& rest, const std::string& url) {
  if (rest.find('@') != std::string::npos)
    throw UnsupportedError("userinfo (user:pass@host) is unsupported — kevy has no AUTH");
  std::string authority = rest;
  std::string path;
  if (size_t slash = rest.find('/'); slash != std::string::npos) {
    authority = rest.substr(0, slash);
    path = rest.substr(slash + 1);
  }
  std::string host = authority;
  uint16_t port = 6379;
  if (size_t colon = authority.rfind(':'); colon != std::string::npos) {
    host = authority.substr(0, colon);
    std::string ps = authority.substr(colon + 1);
    uint32_t p = 0;
    if (!parse_u32(ps, p) || p > 65535) throw InvalidInputError("bad port: " + ps);
    port = static_cast<uint16_t>(p);
  }
  if (host.empty()) throw InvalidInputError("empty host");
  Target t;
  t.kind = TargetKind::Remote;
  t.host = host;
  t.port = port;
  t.url = url;
  // tcp:// is raw: it never carries a db (ignores any /db, does no SELECT).
  if (scheme != "tcp" && !path.empty()) {
    uint32_t n = 0;
    if (!parse_u32(path, n))
      throw InvalidInputError("bad db index: '" + path + "' (expected a non-negative integer)");
    t.db = n;
  }
  return t;
}

}  // namespace

Target parse_connect_url(const std::string& url) {
  size_t sep = url.find("://");
  if (sep == std::string::npos) throw InvalidInputError("URL missing '://'");
  std::string scheme = url.substr(0, sep);
  std::string rest = url.substr(sep + 3);
  if (scheme == "mem") {
    Target t;
    t.url = url;
    if (rest.empty()) {
      t.kind = TargetKind::MemAnon;
    } else {
      t.kind = TargetKind::MemNamed;
      t.name = rest;
    }
    return t;
  }
  if (scheme == "file") {
    if (rest.empty())
      throw InvalidInputError("file:// URL must include a path (e.g. file:///var/lib/myapp)");
    Target t;
    t.kind = TargetKind::File;
    t.path = rest;
    t.url = url;
    return t;
  }
  if (scheme == "kevy" || scheme == "redis" || scheme == "tcp") {
    return parse_remote(scheme, rest, url);
  }
  if (scheme == "rediss" || scheme == "kevys") {
    throw UnsupportedError("TLS schemes (rediss://, kevys://) are unsupported — kevy has no TLS");
  }
  throw InvalidInputError("unknown URL scheme '" + scheme + "://'");
}

// --- process-global embedded registry (§1.3) ----------------------------
//
// A URL-keyed weak map: two connects with the same mem://<name> or
// file://<path> share one store; the entry evicts when the last strong
// (shared_ptr) handle drops.

namespace {
std::mutex g_registry_mu;
std::unordered_map<std::string, std::weak_ptr<EmbeddedStore>> g_registry;

std::shared_ptr<EmbeddedStore> open_embedded(const Target& t) {
  switch (t.kind) {
    case TargetKind::MemAnon:
    case TargetKind::MemNamed:
      return std::make_shared<EmbeddedStore>(EmbeddedStore::open_mem());
    case TargetKind::File:
      return std::make_shared<EmbeddedStore>(EmbeddedStore::open(t.path));
    default:
      throw InvalidInputError("resolve_store called on a non-embedded target");
  }
}
}  // namespace

std::shared_ptr<EmbeddedStore> resolve_store(const Target& t) {
  std::string key = t.registry_key();
  if (key.empty()) return open_embedded(t);  // anonymous mem:// — never shared
  std::lock_guard<std::mutex> lock(g_registry_mu);
  if (auto it = g_registry.find(key); it != g_registry.end()) {
    if (auto strong = it->second.lock()) return strong;
  }
  auto store = open_embedded(t);
  g_registry[key] = store;
  return store;
}

}  // namespace detail
}  // namespace kevy
